docs/pallas/gpu/pipelining.md

```diff
@@ -0,0 +1,325 @@
+<!-- #endregion -->
+
+```python id="6Vf5_VA9iCD1"
+def matmul(a, b, tile_m=128, tile_n=128, swizzle=128):
+  dtype = jnp.float16
+  elems_128b = swizzle // jnp.dtype(dtype).itemsize
+  tile_k = elems_128b
+  grid_m = m // tile_m
+  grid_k = k // tile_k
+  grid_n = n // tile_n
+  assert tile_m % elems_128b == 0
+
+  # Note: Transforms will be inferred automatically
+  # by Mosaic GPU in the future.
+  transforms = (
+          plgpu.TilingTransform((8, elems_128b)),
+          plgpu.SwizzleTransform(128),
+      )
+
+  def kernel(a_gmem, b_gmem, o_gmem, o_smem, acc):
+    def kernel_body(_, a_smem, b_smem):
+      plgpu.wgmma(acc, a_smem, b_smem)
+      plgpu.wgmma_wait(1)
+
+    # pl.program_id obtains the index into the pallas_call grid.
+    pid_m = pl.program_id(0)
+    pid_n = pl.program_id(1)
+
+    pipeline = plgpu.emit_pipeline(
+        kernel_body,
+        in_specs=[
+            plgpu.GPUBlockSpec(
+                (tile_m, tile_k), lambda k: (pid_m, k), transforms=transforms
+            ),
+            plgpu.GPUBlockSpec(
+                (tile_k, tile_n), lambda k: (k, pid_n), transforms=transforms
+            ),
+        ],
+        grid=(grid_k,),
+        max_concurrent_steps=2,
+        delay_release=1,
+    )
+
+    pipeline(a_gmem, b_gmem)
+    # Store WGMMA accumulator to SMEM and then to GMEM.
+    o_smem[...] = acc[...].astype(dtype)
+    plgpu.commit_smem()
+    m_slice = pl.ds(pid_m * tile_m, tile_m)
+    n_slice = pl.ds(pid_n * tile_n, tile_n)
+    plgpu.copy_smem_to_gmem(o_smem, o_gmem.at[m_slice, n_slice])
+    plgpu.wait_smem_to_gmem(0)
+
+  return plgpu.kernel(
+      kernel,
+      out_shape=jax.ShapeDtypeStruct((m, n), jnp.float16),
+      scratch_shapes=[
+          plgpu.SMEM((tile_m, tile_n), jnp.float16),
+          plgpu.ACC((tile_m, tile_n), jnp.float32)
+          ],
+      # grid specifies the CUDA grid.
+      grid=(grid_m, grid_n),
+      grid_names=("m", "n"),
+  )(a, b)
+
+m = 132 * 128
+n = 4 * 128
+k = 10 * 64
+key1, key2 = jax.random.split(jax.random.key(42), 2)
+a = jax.random.uniform(key1, shape=(m, k), dtype=jnp.float16)
+b = jax.random.uniform(key2, shape=(k, n), dtype=jnp.float16)
+
+result = matmul(a, b)
+
+np.testing.assert_allclose(result, a @ b)
+```
+
+<!-- #region id="lIYV7PN9J8Px" -->
+### Warp Specialization
+
+Warp specialization is a technique where we program each warp/warpgroup to perform a single task in order to give the GPU hardware the flexibility to schedule them at runtime. Recall that each streaming multiprocessor (SM) in a GPU contains warp schedulers that can swap execution between warps, so for example when one warp is blocked it can begin executing a different warp. In practice, this can be more performant than programming a single instruction stream and having the compiler statically schedule the operations to overlap the instructions them optimally.
```

Comment:
to _dynamically_ schedule them

Reply:
In the non-warp specialized cast it's not dynamically scheduled right? It's scheduled by the PTX compiler.